bundles/org.openhab.binding.modbus.e3dc/src/main/java/org/openhab/binding/modbus/e3dc/internal/modbus/Parser.java

```diff
@@ -0,0 +1,130 @@
+import org.openhab.binding.modbus.e3dc.internal.dto.InfoBlock;
+import org.openhab.binding.modbus.e3dc.internal.dto.PowerBlock;
+import org.openhab.binding.modbus.e3dc.internal.dto.StringBlock;
+import org.openhab.binding.modbus.e3dc.internal.dto.WallboxArray;
+import org.openhab.binding.modbus.e3dc.internal.modbus.Data.DataType;
+import org.openhab.io.transport.modbus.AsyncModbusReadResult;
+import org.openhab.io.transport.modbus.ModbusRegister;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link InfoBlockCallback} class receives callbacks from modbus poller
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class Parser {
+    private static final int MEASURE_COUNT = 100;
+    private final Logger logger = LoggerFactory.getLogger(Parser.class);
+    private DataType callbackType;
+    private byte[] bArray;
+    private int size;
+    private int counter = 0;
+    private long maxDuration = Long.MIN_VALUE;
+    private long minDuration = Long.MAX_VALUE;
+    private long avgDuration = 0;
+
+    public Parser(DataType type) {
+        callbackType = type;
+        if (type.equals(DataType.INFO)) {
+            size = INFO_REG_SIZE * 2;
+            bArray = new byte[size];
+        } else {
+            size = (REGISTER_LENGTH - INFO_REG_SIZE) * 2;
+            bArray = new byte[size];
+        }
+    }
+
+    public void handle(AsyncModbusReadResult result) {
+        byte[] newArray = new byte[size];
+        long startTime = System.currentTimeMillis();
+        Optional<ModbusRegisterArray> opt = result.getRegisters();
+        if (opt.isPresent()) {
+            ModbusRegisterArray registers = opt.get();
+            int i = 0;
+            for (ModbusRegister reg : registers) {
+                System.arraycopy(reg.getBytes(), 0, newArray, i, 2);
+                i += 2;
+            }
+            setArray(newArray);
+
+            long duration = System.currentTimeMillis() - startTime;
+            avgDuration += duration;
+            minDuration = Math.min(minDuration, duration);
+            maxDuration = Math.max(maxDuration, duration);
+            counter++;
+            if (counter % MEASURE_COUNT == 0) {
+                logger.debug("Min {} Max {} Avg {}", minDuration, maxDuration, avgDuration / MEASURE_COUNT);
+                avgDuration = 0;
+                minDuration = Long.MAX_VALUE;
+                maxDuration = Long.MIN_VALUE;
+            }
+        } else {
+            logger.warn("Modbus read result doesn't return expected registers");
+        }
+    }
+
+    public synchronized void setArray(byte[] b) {
+        if (b.length != size) {
+            logger.warn("Wrong byte size received. Should be {} but is {}. Data maybe corrupted!", size, b.length);
+        }
+        bArray = b.clone();
+    }
+
+    public @Nullable Data parse(DataType type) {
```

Comment:
Consider using Optional<Data>. Matter of taste I guess

Reply:
No problem. To be honest I didn't know the Optional pattern before but it's quite nice to handle.
Working on this topic I observed that E3DCThingHandler creates the channelUIDs everytime on update. Seems that there are heavy String operations each time. I moved Channel creation towards the constructor - more code but should be more efficient. Please have a look on E3DCThingHandler

Reply:
Yeah, it probably does make a difference since there are so many channels. 

Btw, behind the scenes (see Basethinghandler) similar thing might be happening with the wall box thing but there are less channels overall

Reply:
You're right. BasicTingHandler creates Channels also on each update - I didn't recognized this. I was just worried to create channels on the frequent Modbus updates and each time Channels are created. I'll correct also WallboxHandler in order to present a consistent implementation.